### PR TITLE
Allow function reference calls across executors

### PR DESCRIPTION
**Pull request by KevinEady**

### CODE REVIEW

File: pol-core/bscript/executor.cpp [hunk 1]
```diff
@@ -2861,9 +2861,40 @@ void Executor::ins_jsr_userfunc( const Instruction& ins, BContinuation* continua
...
     rc.Continuation.set( new BObject( continuation ) );
   }
+
+  // Only store our global context if the function is external to the current program.
+  if ( funcref != nullptr && funcref->prog() != prog_ )
+  {
+    // Store external context for the return path.
+    rc.ExternalContext = ReturnContext::External( prog_, std::move( execmodules ), Globals2 );
+
+    // Set the prog and globals (if required) to the external function's,
```
Comment (KevinEady) [inline pol-core/bscript/executor.cpp:2871]: Remove `(if required)`

File: pol-core/bscript/executor.cpp [hunk 2]
```diff
@@ -3181,8 +3220,8 @@ void Executor::ins_funcref( const Instruction& ins )
   const auto& ep_funcref = prog_->function_references[funcref_index];
 
   ValueStack.push_back( BObjectRef(
-      new BObject( new BFunctionRef( ins.token.lval, ep_funcref.parameter_count, scriptname(),
-                                     ep_funcref.is_variadic, {} /* captures */ ) ) ) );
+      new BObject( new BFunctionRef( prog_, ins.token.lval, ep_funcref.parameter_count,
```
Comment (KevinEady) [inline pol-core/bscript/executor.cpp:3223]: simplify

**Review by turleypol — DISMISSED** [review]

File: pol-core/bscript/object.cpp [hunk 3]
```diff
@@ -2304,19 +2304,21 @@ std::string BBoolean::getStringRep() const
 }
 
 
-BFunctionRef::BFunctionRef( int progcounter, int param_count, const std::string& scriptname,
-                            bool variadic, ValueStackCont&& captures )
+BFunctionRef::BFunctionRef( ref_ptr<EScriptProgram> program, int progcounter, int param_count,
+                            bool variadic, std::shared_ptr<ValueStackCont> globals,
+                            ValueStackCont&& captures )
```
Comment (turleypol) [inline pol-core/bscript/object.cpp:2309]: Same here: pass by value then move

File: pol-core/bscript/executor.cpp [hunk 4]
```diff
@@ -2928,6 +2957,17 @@ void Executor::ins_return( const Instruction& /*ins*/ )
     upperLocals2.pop_back();
   }
 
+  if ( rc.ExternalContext.has_value() )
+  {
+    prog_ = rc.ExternalContext->Program;
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:2962]: Since rc gets destroyed the members can be moved

File: pol-core/bscript/executor.cpp [hunk 5]
```diff
@@ -193,6 +193,7 @@ Executor::Executor()
       run_ok_( false ),
       debug_level( NONE ),
       PC( 0 ),
+      Globals2( new BObjectRefVec ),
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:196]:
```suggestion
      Globals2(std::make_shared<BObjectRefVec>()),
```

File: pol-core/bscript/executor.h [hunk 6]
```diff
@@ -78,9 +80,23 @@ class ExecutorDebugListener
 // FIXME: how to make this a nested struct in Executor?
 struct ReturnContext
 {
+  struct External
+  {
+    External( ref_ptr<EScriptProgram> program, std::vector<ExecutorModule*> modules,
+              std::shared_ptr<BObjectRefVec> globals )
+        : Program( program ), Modules( modules ), Globals( globals )
```
Comment (turleypol) [inline pol-core/bscript/executor.h:87]:
```suggestion
        : Program( std::move(program) ), Modules( std::move(modules) ), Globals( std::move(globals) )
```

File: pol-core/bscript/object.cpp [hunk 7]
```diff
@@ -2304,19 +2304,21 @@ std::string BBoolean::getStringRep() const
...
     : BObjectImp( OTFuncRef ),
+      prog_( program ),
```
Comment (turleypol) [inline pol-core/bscript/object.cpp:2311]:
```suggestion
      prog_( std::move(program) ),
```

File: pol-core/bscript/object.cpp [hunk 8]
```diff
@@ -2304,19 +2304,21 @@ std::string BBoolean::getStringRep() const
...
       pc_( progcounter ),
       num_params_( param_count ),
-      script_name_( scriptname ),
       variadic_( variadic ),
+      globals( globals ),
```
Comment (turleypol) [inline pol-core/bscript/object.cpp:2315]:
```suggestion
      globals( std::move(globals) ),
```

File: pol-core/bscript/executor.cpp (hunk 4 above)
Comment (turleypol) [inline pol-core/bscript/executor.cpp:2962]:
```suggestion
    prog_ = std::move(rc.ExternalContext->Program);
```

File: pol-core/bscript/executor.cpp [hunk 9]
```diff
@@ -2928,6 +2957,17 @@ void Executor::ins_return( const Instruction& /*ins*/ )
     upperLocals2.pop_back();
   }
 
+  if ( rc.ExternalContext.has_value() )
+  {
+    prog_ = rc.ExternalContext->Program;
+    nLines = static_cast<unsigned int>( prog_->instr.size() );
+    execmodules = rc.ExternalContext->Modules;
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:2964]:
```suggestion
    execmodules = std::move(rc.ExternalContext->Modules);
```

File: pol-core/bscript/executor.cpp [hunk 10]
```diff
@@ -2928,6 +2957,17 @@ void Executor::ins_return( const Instruction& /*ins*/ )
...
+    Globals2 = rc.ExternalContext->Globals;
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:2965]:
```suggestion
    Globals2 = std::move(rc.ExternalContext->Globals);
```

File: pol-core/bscript/executor.cpp [hunk 11]
```diff
@@ -2832,9 +2833,40 @@ void Executor::ins_jsr_userfunc( const Instruction& ins, BContinuation* continua
...
+  if ( funcref != nullptr && funcref->prog() != prog_ )
+  {
+    // Store external context for the return path.
+    rc.ExternalContext = ReturnContext::External( prog_, std::move( execmodules ), Globals2 );
+
+    // Set the prog and globals to the external function's, updating nLines and
+    // execmodules.
+    prog_ = funcref->prog();
+
+    // TODO future improvement: do not copy the globals, but instead point our
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:2847]: This Todo is with the change to sharedptr done I would say

**Review by turleypol — APPROVED** [review]

**Pull request merged**